src/Symfony/Bundle/FrameworkBundle/Tests/Controller/ControllerTraitTest.php

```diff
@@ -0,0 +1,388 @@
+    }
+
+    public function testStreamTwig()
+    {
+        $twig = $this->getMockBuilder(\Twig_Environment::class)->disableOriginalConstructor()->getMock();
+
+        $controller = new TestTrait();
+        $controller->setTwig($twig);
+
+        $this->assertInstanceOf(StreamedResponse::class, $controller->stream('foo'));
+    }
+
+    public function testCreateNotFoundException()
+    {
+        $controller = new TestTrait();
+
+        $this->assertInstanceOf(NotFoundHttpException::class, $controller->createNotFoundException());
+    }
+
+    public function testCreateAccessDeniedException()
+    {
+        $controller = new TestTrait();
+
+        $this->assertInstanceOf(AccessDeniedException::class, $controller->createAccessDeniedException());
+    }
+
+    public function testCreateForm()
+    {
+        $form = $this->getMock(FormInterface::class);
+
+        $formFactory = $this->getMock(FormFactoryInterface::class);
+        $formFactory->expects($this->once())->method('create')->willReturn($form);
+
+        $controller = new TestTrait();
+        $controller->setFormFactory($formFactory);
+
+        $this->assertEquals($form, $controller->createForm('foo'));
+    }
+
+    public function testCreateFormBuilder()
+    {
+        $formBuilder = $this->getMock(FormBuilderInterface::class);
+
+        $formFactory = $this->getMock(FormFactoryInterface::class);
+        $formFactory->expects($this->once())->method('createBuilder')->willReturn($formBuilder);
+
+        $controller = new TestTrait();
+        $controller->setFormFactory($formFactory);
+
+        $this->assertEquals($formBuilder, $controller->createFormBuilder('foo'));
+    }
+
+    public function testGetDoctrine()
+    {
+        $doctrine = $this->getMock(ManagerRegistry::class);
+
+        $controller = new TestTrait();
+        $controller->setDoctrine($doctrine);
+
+        $this->assertEquals($doctrine, $controller->getDoctrine());
+    }
+
+    public function testIsCsrfTokenValid()
+    {
+        $tokenManager = $this->getMock(CsrfTokenManagerInterface::class);
+        $tokenManager->expects($this->once())->method('isTokenValid')->willReturn(true);
+
+        $controller = new TestTrait();
+        $controller->setCsrfTokenManager($tokenManager);
+
+        $this->assertTrue($controller->isCsrfTokenValid('foo', 'bar'));
+    }
+}
+
+class TestTrait
+{
+    use ControllerTrait {
+        forward as public;
+        getUser as public;
+        json as public;
+        generateUrl as public;
+        redirect as public;
+        redirectToRoute as public;
+        addFlash as public;
+        isGranted as public;
+        denyAccessUnlessGranted as public;
+        renderView as public;
+        render as public;
+        stream as public;
+        createNotFoundException as public;
+        createAccessDeniedException as public;
+        createForm as public;
+        createFormBuilder as public;
+        getDoctrine as public;
+        isCsrfTokenValid as public;
```

Comment:
As explained in the description, I think this is useless. They are proxy methods to the service itself. If you don't want the convenience of this "all-in-one" trait, just inject the service and use it.